sqlglot/optimizer/qualify_columns.py

```diff
@@ -380,18 +410,105 @@ def _qualify_columns(scope: Scope, resolver: Resolver) -> None:
                     column.set("table", column_table)
 
 
+def _expand_struct_stars(
+    expression: exp.Dot,
+    next_alias_name: t.Callable[[], str],
+) -> t.List[exp.Alias]:
+    """[BigQuery] Expand/Flatten foo.bar.* where bar is a struct column"""
+
+    dot_column = t.cast(exp.Column, expression.find(exp.Column))
+    if not dot_column.is_type(exp.DataType.Type.STRUCT):
+        return []
+
+    # All nested struct values are ColumnDefs, so normalize the first exp.Column in one
+    dot_column = dot_column.copy()
+    starting_struct = exp.ColumnDef(this=dot_column.this, kind=dot_column.type)
+
+    # First part is the table name and last part is the star so they can be dropped
+    dot_parts = expression.parts[1:-1]
+
+    # If we're expanding a nested struct (eg. t.c.f1.f2.*), start the DFS traversal from that struct
+    for part in dot_parts[1:]:
+        struct_fields = t.cast(exp.DataType, starting_struct.kind).expressions
+
+        if not struct_fields:
+            return []
+
+        for fld in struct_fields:
+            # Unable to expand star unless all struct fields are named
+            if not isinstance(fld.this, exp.Identifier):
+                return []
+
+            if fld.name == part.name and fld.kind.is_type(exp.DataType.Type.STRUCT):
+                starting_struct = fld
```

Comment:
Perhaps something interesting to notice here is that if, for whatever reason, there is no `fld` that matches the current `part`, we'll keep traversing `dot_parts` using the same `starting_struct`. What happens if there _is_ a matching `fld` for the *next* dot part? It looks like we'll happily accept that and update the `starting_struct`, which may lead to an invalid expansion later on.

Reply:
Interesting take. Not only that, but looking at the nested loops I've made it so the internal loop keeps searching even if it has found the next `fld` already.

In my follow up commit I'll fix both issues at once by checking if each part has matched; If yes, we `break` and go to the next part, otherwise we fail early and return `[]`

Reply:
I was thinking perhaps we can get rid of the condition below, where we check if the starting struct matches the last dot part. I think one invariant in this loop is that for each element in `dot_parts`, we must find at least one `fld` such that `fld.name == part.name`, otherwise there's no progress made. So if this invariant is true by the end of the loop, that condition would be redundant.

Reply:
This suggestion was implemented in feedback n.6